File: frontend/src/api/meetings.ts
import { api } from "./client";

export interface MeetingAttendee {
  user_id: number;
  rsvp: number; // 0=нет ответа, 1=да, 2=может, 3=нет
}

export interface MeetingDetail {
  id: number;
  created_by: number;
  title: string;
  starts_at: string;
  ends_at: string;
  location: string | null;
  status: string;
  auto_picked: boolean;
  score: number | null;
  attendees: MeetingAttendee[];
  my_rsvp: number;
}

export const fetchMeetings = (from: Date, to: Date) =>
  api<MeetingDetail[]>(
    `/api/meetings?from=${from.toISOString()}&to=${to.toISOString()}`,
  );

export const setRsvp = (meetingId: number, rsvp: number) =>
  api<MeetingDetail>(`/api/meetings/${meetingId}/rsvp`, {
    method: "PATCH",
    body: JSON.stringify({ rsvp }),
  });

export const cancelMeeting = (meetingId: number) =>
  api<void>(`/api/meetings/${meetingId}`, { method: "DELETE" });

export interface AutoPickSlot {
  starts_at: string;
  ends_at: string;
  score: number;
  available_user_ids: number[];
  maybe_user_ids: number[];
}

export interface AutoPickResponse {
  slots: AutoPickSlot[];
}

export interface AutoPickRequest {
  window_start: string;
  window_end: string;
  duration_minutes: number;
  step_minutes: number;
  top_n: number;
}

export const autoPick = (body: AutoPickRequest) =>
  api<AutoPickResponse>("/api/meetings/auto-pick", {
    method: "POST",
    body: JSON.stringify(body),
  });

export interface LoserRoll {
  id: number;
  rolled_at: string;
  rolled_by: number;
  loser_user_id: number;
  reason_text: string | null;
}

export interface LoserRollResponse {
  roll: LoserRoll;
  sent_to_chat: boolean;
}

export interface LoserStats {
  counts: Record<number, number>;
  last: LoserRoll | null;
  cooldown_remaining_seconds: number;
}

export const rollLoser = () =>
  api<LoserRollResponse>("/api/loser/roll", { method: "POST" });

export const fetchLoserStats = () => api<LoserStats>("/api/loser/stats");

export const fetchLoserHistory = (limit = 20) =>
  api<LoserRoll[]>(`/api/loser/history?limit=${limit}`);

export interface PollOption {
  id: number;
  starts_at: string;
  label: string | null;
  voter_user_ids: number[];
}

export interface Poll {
  id: number;
  question: string;
  closes_at: string | null;
  options: PollOption[];
  my_vote_option_id: number | null;
}

export interface PollCreateRequest {
  question: string;
  options: string[]; // ISO datetimes
  closes_in_hours?: number | null;
  chat_id?: number | null;
}

export const createPoll = (body: PollCreateRequest) =>
  api<Poll>("/api/polls", {
    method: "POST",
    body: JSON.stringify(body),
  });

export const fetchPolls = () => api<Poll[]>("/api/polls");

export interface PollAutoPickRequest {
  window_start: string;
  window_end: string;
  duration_minutes: number;
  step_minutes: number;
  top_n: number;
  question?: string;
  closes_in_hours?: number | null;
  chat_id?: number | null;
}

export const createAutoPickPoll = (body: PollAutoPickRequest) =>
  api<Poll>("/api/polls/auto-pick", {
    method: "POST",
    body: JSON.stringify(body),
  });

export interface IcalUrl {
  https: string;
  webcal: string;
}

export const fetchIcalUrl = () => api<IcalUrl>("/api/meetings/ical/url");

File: backend/app/schemas/meetings.py
from __future__ import annotations

from datetime import datetime, timedelta

from pydantic import BaseModel, ConfigDict, Field


class AutoPickRequest(BaseModel):
    """Request body for /api/meetings/auto-pick."""

    window_start: datetime
    window_end: datetime
    duration_minutes: int = Field(120, ge=30, le=24 * 60)
    step_minutes: int = Field(60, ge=15, le=24 * 60)
    top_n: int = Field(5, ge=1, le=20)
    # GHG5 POLL-HOURS1: использовать пресеты времени из admin_config вместо
    # фиксированных duration/step. True по умолчанию — никаких больше 00-04.
    use_presets: bool = True


class AutoPickSlotOut(BaseModel):
    starts_at: datetime
    ends_at: datetime
    score: float
    available_user_ids: list[int]
    maybe_user_ids: list[int]


class AutoPickResponse(BaseModel):
    slots: list[AutoPickSlotOut]


class MeetingCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    starts_at: datetime
    ends_at: datetime
    location: str | None = None
    auto_picked: bool = False
    score: float | None = None


class MeetingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_by: int
    title: str
    starts_at: datetime
    ends_at: datetime
    location: str | None = None
    status: str
    auto_picked: bool
    score: float | None = None


class RsvpRequest(BaseModel):
    rsvp: int = Field(..., ge=0, le=3)  # 0=не ответил, 1=да, 2=может, 3=нет


class MeetingAttendeeOut(BaseModel):
    user_id: int
    rsvp: int


class MeetingDetail(MeetingOut):
    attendees: list[MeetingAttendeeOut]
    my_rsvp: int = 0


class LoserRollOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    rolled_at: datetime
    rolled_by: int
    loser_user_id: int
    reason_text: str | None = None


class LoserStatsOut(BaseModel):
    counts: dict[int, int]
    last: LoserRollOut | None
    cooldown_remaining_seconds: int


class LoserRollResponse(BaseModel):
    roll: LoserRollOut
    sent_to_chat: bool


class PollCreateRequest(BaseModel):
    """GHG6 PL5: вариант может быть либо `datetime` ISO, либо строка-дата `YYYY-MM-DD`.

    Date-only означает, что время не задано (опросник без чекбокса «указать время»):
    backend хранит `starts_at = date 00:00 локали`, `ends_at = 23:59`, label
    «вс 17.05» — без часа.
    """

    question: str = Field(..., min_length=1, max_length=255)
    # Принимаем строки (ISO datetime или YYYY-MM-DD). Парсинг — в services/polls.py.
    options: list[str] = Field(..., min_length=2, max_length=6)
    closes_in_hours: int | None = Field(None, ge=1, le=72)
    chat_id: int | None = None
    # G2: закрепить сообщение с опросом. None → дефолт из admin_config.
    pin: bool | None = None


class PollAutoPickRequest(BaseModel):
    """Авто-публикация TG-опроса по найденным top-N слотам."""

    window_start: datetime
    window_end: datetime
    duration_minutes: int = Field(120, ge=30, le=24 * 60)
    step_minutes: int = Field(60, ge=15, le=24 * 60)
    top_n: int = Field(3, ge=2, le=5)
    question: str = Field("Когда соберёмся?", min_length=1, max_length=255)
    closes_in_hours: int | None = Field(24, ge=1, le=72)
    chat_id: int | None = None
    # GHG5 POLL-HOURS1: использовать пресеты времени из admin_config.
    use_presets: bool = True
    # G2: закрепить сообщение. None → дефолт из admin_config.
    pin: bool | None = None


class PollOptionOut(BaseModel):
    id: int
    starts_at: datetime
    label: str | None = None
    voter_user_ids: list[int] = []


class PollOut(BaseModel):
    id: int
    question: str
    closes_at: datetime | None
    options: list[PollOptionOut]
    my_vote_option_id: int | None = None


def fmt_remaining(td: timedelta) -> str:
    total = int(td.total_seconds())
    h, rem = divmod(total, 3600)
    m, _ = divmod(rem, 60)
    return f"{h}ч {m}мин" if h else f"{m}мин"
